### PR TITLE
Ticketvalidation: UX Show Validation Result (EXPOSUREAPP-10196)

### PR DESCRIPTION
**Pull request by jurajkusnier**
### Ticketvalidation: UX Show Validation Result

out
out_black

### Jira Link
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10196

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/DccTicketingValidationResultFragment.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    @Inject lateinit var viewModelFactory: CWAViewModelFactoryProvider.Factory
+    @Inject lateinit var validationResultAdapter: ValidationResultAdapter
+
+    private val binding: FragmentDccTicketingValidationResultBinding by viewBinding()
+    private val args: DccTicketingValidationResultFragmentArgs by navArgs()
+    private val resultViewModel: DccTicketingValidationResultViewModel by cwaViewModelsAssisted(
+        factoryProducer = { viewModelFactory },
+        constructorCall = { factory, _ ->
+            factory as DccTicketingValidationResultViewModel.Factory
+            factory.create(transactionContext = args.transactionContext)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/DccTicketingValidationResultFragment.kt:30]: Please use the sharedViewModel to get the latest TransactionContext see https://github.com/corona-warn-app/cwa-app-android/pull/4439#discussion_r757413627

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/ValidationResultAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    init {
+        listOf(
+            StableIdMod(data),
+            DataBinderMod<ValidationResultItem, BaseValidationResultVH<ValidationResultItem, ViewBinding>>(data),
+            TypedVHCreatorMod({ data[it] is ValidationFaqVH.Item }) { ValidationFaqVH(it) },
+            TypedVHCreatorMod({ data[it] is DescriptionVH.Item }) { DescriptionVH(it) },
+            TypedVHCreatorMod({ data[it] is TestingInfoVH.Item }) { TestingInfoVH(it) },
+            TypedVHCreatorMod({ data[it] is ResultRuleVH.Item }) { ResultRuleVH(it) },
+            // TODO: more items here
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/ValidationResultAdapter.kt:32]: Do you expect more items to be added ? or ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/DccTicketingValidationResultFragment.kt (hunk 1 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/DccTicketingValidationResultFragment.kt:30]: Thanks, yes, I'm aware of this. This is actually the reason why the PR is still in draft mode.

Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/DccTicketingValidationResultFragment.kt]: updated

**Review by KathrinLuecke — DISMISSED** [review]
UA approved

File: Corona-Warn-App/src/main/res/navigation/dcc_ticketing_nav_graph.xml [hunk 3]
```diff
@@ -49,7 +49,7 @@
     </fragment>
     <fragment
         android:id="@+id/dccTicketingValidationSuccessFragment"
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/navigation/dcc_ticketing_nav_graph.xml:51]: Please rename to `dccTicketingValidationResultFragment`?

File: Corona-Warn-App/src/main/res/navigation/dcc_ticketing_nav_graph.xml [hunk 4]
```diff
@@ -49,7 +49,7 @@
...
-        android:name="de.rki.coronawarnapp.dccticketing.ui.validationresult.failed.DccTicketingValidationFailedFragment"
-        android:label="fragment_dcc_ticketing_validation_failed"
-        tools:layout="@layout/fragment_dcc_ticketing_validation_failed" />
+        android:name="de.rki.coronawarnapp.dccticketing.ui.validationresult.DccTicketingValidationResultFragment"
+        android:label="fragment_dcc_ticketing_validation_success"
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/navigation/dcc_ticketing_nav_graph.xml:53]: `.._result`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/DccTicketingValidationResultFragment.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+        resultViewModel.navigation.observe2(this) {
+            handleNavigation(it)
+        }
+    }
+
+    private fun handleNavigation(navigation: DccTicketingValidationNavigation) {
+        when (navigation) {
+            DccTicketingValidationNavigation.Close -> popBackStack()
+            DccTicketingValidationNavigation.Done -> popBackStack()
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/DccTicketingValidationResultFragment.kt:69]:
```
DccTicketingValidationNavigation.Close ,
DccTicketingValidationNavigation.Done -> popBackStack()
```

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/success/DccTicketingValidationResultFragmentTest.kt [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        )
+    }
+
+    private val navController = TestNavHostController(
+        ApplicationProvider.getApplicationContext()
+    ).apply {
+        UiThreadStatement.runOnUiThread {
+            setViewModelStore(ViewModelStore())
+            setGraph(R.navigation.nav_graph)
+            setCurrentDestination(R.id.dccTicketingValidationSuccessFragment)
```
Comment (mtwalli) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/success/DccTicketingValidationResultFragmentTest.kt:56]: `dccTicketingValidationResultFragment`?

**Review by mtwalli — COMMENTED** [review]
LGTM, just minor stuff

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/success/DccTicketingTransactionContextData.kt [hunk 7]
```diff
@@ -0,0 +1,78 @@
+package de.rki.coronawarnapp.dccticketing.ui.validationresult.success
+
+import de.rki.coronawarnapp.dccticketing.core.qrcode.DccTicketingQrCodeData
+import de.rki.coronawarnapp.dccticketing.core.transaction.DccTicketingResultItem
+import de.rki.coronawarnapp.dccticketing.core.transaction.DccTicketingResultToken
+import de.rki.coronawarnapp.dccticketing.core.transaction.DccTicketingTransactionContext
+
+val dccTicketingTransactionContextPassed = DccTicketingTransactionContext(
```
Comment (BMItr) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/success/DccTicketingTransactionContextData.kt:8]: You could reduce the code by just using one DccTicketingTransactionContext and change the values of result and results with copy..

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/DccTicketingValidationResultViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            result = context.resultTokenPayload?.result ?: throw NullPointerException("resultTokenPayload is null"),
+            listItems = generateItems(context.resultTokenPayload, context.initializationData.serviceProvider)
+        )
+    }.asLiveData2()
+
+    private fun generateItems(
+        resultToken: DccTicketingResultToken,
+        serviceProvider: String
+    ): List<ValidationResultItem> = with(itemCreator) {
+        mutableListOf(
```
Comment (BMItr) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/validationresult/DccTicketingValidationResultViewModel.kt:36]: This should be also part of item creator, i guess

**Review by mtwalli — APPROVED** [review]
LGTM

**Review by BMItr — APPROVED** [review]
lgtm!

**Review by AlexanderAlferov — APPROVED** [review]

**Comment by vaubaehn** [comment]
@KathrinLuecke @mtwalli 
What is the difference between "Zertifikat nicht prüfbar" (status "open"?) and "Zertifikat nicht anerkannt" (status "failed")?
The explaining text "Ihr Zertifikat entspricht nicht den Anforderungen des Anbieters" is same for both - shouldn't be the text for "open" status "Ihr Zertifikat konnte vom Anbieter nicht überprüft werden"?
Using a (presumably) expired RAT for "open" status for the unit test/screen shot is a bit irritating, as an expired test should have status "failed"?
At least the output is not consistent with the "known" output from business rule validation.

**Comment by KathrinLuecke** [comment]
@vaubaehn @mtwalli 
The screen "Zertifikat nicht prüfbar" comes up when the certificate could not be validated at all. That's the reason why for an expired RAT, the status is not failed, becaues an expired RAT is not validated at all. 
But maybe the sentence "Ihr Zertifikat entspricht nicht den Andorderungen des Anbieters" is a bit misleading then.
I will check this.

**Comment by vaubaehn** [comment]
@KathrinLuecke @mtwalli and also inviting @mlenkeit 
I'm referring to the official DCC ticketing manual https://ec.europa.eu/health/sites/default/files/ehealth/docs/covid-certificate_traveller-onlinebooking_en.pdf for all my following evaluations.
> The screen "Zertifikat nicht prüfbar" comes up when the certificate could not be validated at all.

This is true, for the certificate as a whole or for a single validation rule. The result of the validation is "open" then. Possible reasons, why a certificate can't be validated (status "open"): technical reasons (i.e., logical, structural or syntax error in the validation rule), or incompatible rule-dcc-versions (e.g., rule requires DCC schema version 1.3.0, DCC was issued in schema 1.0.0).
However, according to the manual referenced above, the status "open" is also assigned to the status "yellow" in the result displayed by the provider, when certain restrictions may take effect. Please see p. 47 of the manual for an example (Manuela König), where a vaccination certificate _might_ be rejected on entry (for unknown reasons). Apparently, this may require certain restriction rules (of type "acceptance" however), see p. 49. I'm not sure whether this restriction rule example is actually evaluating as "open", however... (should evaluate as "null" to be "open" from my understanding).

> That's the reason why for an expired RAT, the status is not failed, becaues an expired RAT is not validated at all.

This is not true. An expired RAT can actually be validated, but the result is "failed". See again p. 47 for an example (Jonas König).
This is why the unit test/the screenshot is misleading here (and can cause confusion in the users when used for explanations e.g. in blog on release), beside the sentence "Ihr Zertifikat entspricht nicht den Anforderungen des Anbieters" for the "open" status.
All the three (unit test/screenshot/text) should be changed for "open" status.

While the issue here is clearly not consistent with the manual, the manual leaves many questions open, imho.

**Pull request merged**